[libc][math][c23] Fix bounds checking and add FE_INVALID raising in {,u}fromfp{,x}*

overmighty opened the pull request:
See
https://github.com/llvm/llvm-project/pull/86692#issuecomment-2024044889
and
https://github.com/llvm/llvm-project/pull/86892#discussion_r1542276037.

cc @lntue @nickdesaulniers

overmighty committed:
[libc][math][c23] Fix bounds checking in {,u}fromfp{,x}*

See
https://github.com/llvm/llvm-project/pull/86692#issuecomment-2024044889.

overmighty committed:
fixup! [libc][math][c23] Fix bounds checking in {,u}fromfp{,x}*

lntue commented on libc/src/__support/FPUtil/NearestIntegerOperations.h:
```diff
@@ -247,6 +247,11 @@ round_using_current_rounding_mode(T x) {
+      FPBits<T>::SIG_MASK - FPBits<T>::FRACTION_MASK;
+
   if (width == 0U)
     return FPBits<T>::quiet_nan().get_val();
```
I just re-read section F.10.6.10 and F.10.6.11 in N3096, it looks like we will need to raise `FE_INVALID` for those out-of-range cases.

overmighty committed:
[libc][math][c23] Fix bounds checking and add FE_INVALID raising in {,u}fromfp{,x}*

See
https://github.com/llvm/llvm-project/pull/86692#issuecomment-2024044889
and
https://github.com/llvm/llvm-project/pull/86892#discussion_r1542276037.
